### PR TITLE
PW-3134 - Klarna - fix for guest shoppers

### PR DESCRIPTION
**Pull request by martinsrenato**
- Always pass customer object so data can be set on PaymentsRequest and passed to Klarna redirect
- Fixed redirect after payment error with guest shopper

**Tested scenarios**
Tested with Klarna and a few other redirect methods, also tested credit cards

### CODE REVIEW

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 1]
```diff
@@ -453,16 +453,16 @@ private void updateOrderPaymentStatusAndInfo(OrderModel orderModel, PaymentsResp
     @Override
     public OrderData authorisePayment(final HttpServletRequest request, final CartData cartData) throws Exception {
         CheckoutCustomerStrategy checkoutCustomerStrategy = getCheckoutCustomerStrategy();
-        CustomerModel customer = null;
-        if (! checkoutCustomerStrategy.isAnonymousCheckout()) {
-            customer = checkoutCustomerStrategy.getCurrentUserForCheckout();
+
+        CustomerModel customer = checkoutCustomerStrategy.getCurrentUserForCheckout();
```
Comment (rkewlani) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:457]: This is changing logic here. If customer is guest customer and guestTokenization is not enabled, then too customer will be instantiated always. Is it intended? There could be some flows which are identifying guest customer by checking if(customer==null) condition.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java [hunk 2]
```diff
@@ -163,7 +163,7 @@ public String placeOrder(@ModelAttribute("placeOrderForm") final PlaceOrderForm
         //Validate the cart
         if (validateCart(redirectModel)) {
             // Invalid cart. Bounce back to the cart page.
-            return REDIRECT_PREFIX + "/cart";
+            return REDIRECT_PREFIX + AdyenControllerConstants.CART_PREFIX;
```
Comment (rkewlani) [inline adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java:166]: We can import AdyenControllerConstants.CART_PREFIX and use CART_PREFIX everywhere.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java [hunk 3]
```diff
@@ -302,7 +302,7 @@ public String authorise3DS2Payment(final Model model,
         }
 
         LOGGER.warn("Redirecting to cart page...");
-        return REDIRECT_PREFIX + "/cart";
+        return REDIRECT_PREFIX + AdyenControllerConstants.CART_PREFIX;
```
Comment (rkewlani) [inline adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java:305]: same here

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 1 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:456]: Talked offline.
Customer must be passed even on gust checkout, so shopper details can be passed on request.
Moved check for recurring contract flow for guest (isGuestUserTokenizationEnabled flag) to a later stage, on factory class.

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java (hunk 2 above)
Comment (martinsrenato) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java]: Done

File: adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java (hunk 3 above)
Comment (martinsrenato) [reply adyenv6b2ccheckoutaddon/acceleratoraddon/web/src/com/adyen/v6/controllers/pages/AdyenSummaryCheckoutStepController.java]: Done

**Review by rkewlani — APPROVED** [review]

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 4]
```diff
@@ -1404,7 +1397,9 @@ private void restoreCartFromOrder(String orderCode) throws CalculationException,
             getCartService().setSessionCart(cartModel);
         }
 
-        if(hasUserContextChanged(orderModel, cartModel)) {
+        Boolean isAnonymousCheckout = getCheckoutCustomerStrategy().isAnonymousCheckout();
```
Comment (achawki) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:1400]: any specific reason for the boxed boolean

**Review by achawki — APPROVED** [review]

**Pull request merged**